File: kernels/ai/runtime/python/src/runtime/domain/contracts.py
"""Contratos canónicos reutilizables para ecosystem AI."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Final

ROUTING_SOURCE_COPILOT_AGENT: Final[str] = "copilot_agent"
ROUTING_SOURCE_ORCHESTRATOR: Final[str] = "orchestrator"
ROUTING_SOURCE_READ_FALLBACK: Final[str] = "read_fallback"
ROUTING_SOURCE_UI_HINT: Final[str] = "ui_hint"

ALL_ROUTING_SOURCES: Final[tuple[str, ...]] = (
    ROUTING_SOURCE_COPILOT_AGENT,
    ROUTING_SOURCE_ORCHESTRATOR,
    ROUTING_SOURCE_READ_FALLBACK,
    ROUTING_SOURCE_UI_HINT,
)

SERVICE_KIND_INSIGHT: Final[str] = "insight_service"
SERVICE_KIND_GOVERNANCE: Final[str] = "governance_service"
SERVICE_KIND_SYNTHESIS: Final[str] = "synthesis_service"
SERVICE_KIND_INTELLIGENCE: Final[str] = "intelligence_service"

ALL_SERVICE_KINDS: Final[tuple[str, ...]] = (
    SERVICE_KIND_INSIGHT,
    SERVICE_KIND_GOVERNANCE,
    SERVICE_KIND_SYNTHESIS,
    SERVICE_KIND_INTELLIGENCE,
)

OUTPUT_KIND_CHAT_REPLY: Final[str] = "chat_reply"
OUTPUT_KIND_COPILOT_EXPLANATION: Final[str] = "copilot_explanation"
OUTPUT_KIND_INSIGHT_NOTIFICATION: Final[str] = "insight_notification"
OUTPUT_KIND_INSIGHT_SUMMARY: Final[str] = "insight_summary"
OUTPUT_KIND_GOVERNANCE_DECISION: Final[str] = "governance_decision"
OUTPUT_KIND_LLM_ARTIFACT: Final[str] = "llm_artifact"
OUTPUT_KIND_INTELLIGENCE_REPORT: Final[str] = "intelligence_report"

ALL_OUTPUT_KINDS: Final[tuple[str, ...]] = (
    OUTPUT_KIND_CHAT_REPLY,
    OUTPUT_KIND_COPILOT_EXPLANATION,
    OUTPUT_KIND_INSIGHT_NOTIFICATION,
    OUTPUT_KIND_INSIGHT_SUMMARY,
    OUTPUT_KIND_GOVERNANCE_DECISION,
    OUTPUT_KIND_LLM_ARTIFACT,
    OUTPUT_KIND_INTELLIGENCE_REPORT,
)

DEFAULT_LANGUAGE_CODE: Final[str] = "es"
LANGUAGE_CODE_ES: Final[str] = "es"
LANGUAGE_CODE_EN: Final[str] = "en"

ALL_LANGUAGE_CODES: Final[tuple[str, ...]] = (
    LANGUAGE_CODE_ES,
    LANGUAGE_CODE_EN,
)


def is_known_routing_source(name: str | None) -> bool:
    return bool(name and name in ALL_ROUTING_SOURCES)


def normalize_routing_source(name: str | None) -> str:
    if is_known_routing_source(name):
        return str(name)
    return ROUTING_SOURCE_ORCHESTRATOR


def normalize_language_code(name: str | None) -> str:
    if not name:
        return DEFAULT_LANGUAGE_CODE
    normalized = str(name).strip().lower()
    if normalized in ALL_LANGUAGE_CODES:
        return normalized
    return DEFAULT_LANGUAGE_CODE


@dataclass(frozen=True)
class AIRequestContext:
    request_id: str | None = None
    tenant_id: str | None = None
    org_id: str | None = None
    user_id: str | None = None
    actor_id: str | None = None
    preferred_language: str | None = None
    content_language: str | None = None
    routed_agent: str | None = None
    routing_source: str | None = None
    service_kind: str | None = None
    output_kind: str | None = None
    policy_profile: str | None = None
    policy_version: str | None = None


@dataclass(frozen=True)
class NotificationChatHandoff:
    notification_id: str | None = None
    title: str | None = None
    body: str | None = None
    scope: str | None = None
    routed_agent: str | None = None
    content_language: str | None = None
    suggested_user_message: str | None = None
    source_notification_kind: str | None = None
    entity_type: str | None = None
    entity_id: str | None = None


def normalize_notification_chat_handoff(
    payload: dict[str, object] | None,
) -> NotificationChatHandoff:
    payload = payload or {}

    def _clean(key: str) -> str | None:
        value = payload.get(key)
        if value is None:
            return None
        text = str(value).strip()
        return text or None

    return NotificationChatHandoff(
        notification_id=_clean("notification_id"),
        title=_clean("title"),
        body=_clean("body"),
        scope=_clean("scope"),
        routed_agent=_clean("routed_agent"),
        content_language=normalize_language_code(_clean("content_language")),
        suggested_user_message=_clean("suggested_user_message"),
        source_notification_kind=_clean("source_notification_kind") or _clean("kind"),
        entity_type=_clean("entity_type"),
        entity_id=_clean("entity_id"),
    )

File: _import/core/ai/go/contracts.go
package ai

// Contratos canónicos reutilizables para ecosystem AI.

import "strings"

const (
	RoutingSourceCopilotAgent = "copilot_agent"
	RoutingSourceOrchestrator = "orchestrator"
	RoutingSourceReadFallback = "read_fallback"
	RoutingSourceUIHint       = "ui_hint"
)

var AllRoutingSources = []string{
	RoutingSourceCopilotAgent,
	RoutingSourceOrchestrator,
	RoutingSourceReadFallback,
	RoutingSourceUIHint,
}

const (
	ServiceKindInsight      = "insight_service"
	ServiceKindGovernance   = "governance_service"
	ServiceKindSynthesis    = "synthesis_service"
	ServiceKindIntelligence = "intelligence_service"
)

var AllServiceKinds = []string{
	ServiceKindInsight,
	ServiceKindGovernance,
	ServiceKindSynthesis,
	ServiceKindIntelligence,
}

const (
	OutputKindChatReply           = "chat_reply"
	OutputKindCopilotExplanation  = "copilot_explanation"
	OutputKindInsightNotification = "insight_notification"
	OutputKindInsightSummary      = "insight_summary"
	OutputKindGovernanceDecision  = "governance_decision"
	OutputKindLLMArtifact         = "llm_artifact"
	OutputKindIntelligenceReport  = "intelligence_report"
)

var AllOutputKinds = []string{
	OutputKindChatReply,
	OutputKindCopilotExplanation,
	OutputKindInsightNotification,
	OutputKindInsightSummary,
	OutputKindGovernanceDecision,
	OutputKindLLMArtifact,
	OutputKindIntelligenceReport,
}

const (
	DefaultLanguageCode = "es"
	LanguageCodeES      = "es"
	LanguageCodeEN      = "en"
)

var AllLanguageCodes = []string{
	LanguageCodeES,
	LanguageCodeEN,
}

func IsKnownRoutingSource(name string) bool {
	for _, item := range AllRoutingSources {
		if item == name {
			return true
		}
	}
	return false
}

func NormalizeRoutingSource(name string) string {
	if IsKnownRoutingSource(name) {
		return name
	}
	return RoutingSourceOrchestrator
}

func NormalizeLanguageCode(name string) string {
	normalized := strings.TrimSpace(strings.ToLower(name))
	for _, item := range AllLanguageCodes {
		if item == normalized {
			return normalized
		}
	}
	return DefaultLanguageCode
}

type RequestContext struct {
	RequestID         string `json:"request_id,omitempty"`
	TenantID          string `json:"tenant_id,omitempty"`
	OrgID             string `json:"org_id,omitempty"`
	UserID            string `json:"user_id,omitempty"`
	ActorID           string `json:"actor_id,omitempty"`
	PreferredLanguage string `json:"preferred_language,omitempty"`
	ContentLanguage   string `json:"content_language,omitempty"`
	RoutedAgent       string `json:"routed_agent,omitempty"`
	RoutingSource     string `json:"routing_source,omitempty"`
	ServiceKind       string `json:"service_kind,omitempty"`
	OutputKind        string `json:"output_kind,omitempty"`
	PolicyProfile     string `json:"policy_profile,omitempty"`
	PolicyVersion     string `json:"policy_version,omitempty"`
}

type NotificationChatHandoff struct {
	NotificationID         string `json:"notification_id,omitempty"`
	Title                  string `json:"title,omitempty"`
	Body                   string `json:"body,omitempty"`
	Scope                  string `json:"scope,omitempty"`
	RoutedAgent            string `json:"routed_agent,omitempty"`
	ContentLanguage        string `json:"content_language,omitempty"`
	SuggestedUserMessage   string `json:"suggested_user_message,omitempty"`
	SourceNotificationKind string `json:"source_notification_kind,omitempty"`
	EntityType             string `json:"entity_type,omitempty"`
	EntityID               string `json:"entity_id,omitempty"`
}

func NormalizeNotificationChatHandoff(h NotificationChatHandoff) NotificationChatHandoff {
	h.NotificationID = strings.TrimSpace(h.NotificationID)
	h.Title = strings.TrimSpace(h.Title)
	h.Body = strings.TrimSpace(h.Body)
	h.Scope = strings.TrimSpace(h.Scope)
	h.RoutedAgent = strings.TrimSpace(h.RoutedAgent)
	h.ContentLanguage = NormalizeLanguageCode(h.ContentLanguage)
	h.SuggestedUserMessage = strings.TrimSpace(h.SuggestedUserMessage)
	h.SourceNotificationKind = strings.TrimSpace(h.SourceNotificationKind)
	h.EntityType = strings.TrimSpace(h.EntityType)
	h.EntityID = strings.TrimSpace(h.EntityID)
	return h
}
